src/triage/component/audition/thresholding.py

```diff
@@ -21,6 +22,63 @@ def _of_metric(df, metric_filter):
     ]
 
 
+def model_groups_filter(
+    train_end_times,
+    initial_model_group_ids,
+    models_table,
+    db_engine
+):
+    """Before creating the distance table, we want to filter the model_group_ids
+    and train_end_times only if they are reasonable -- every model group should
+    have the same train_end_times -- to prevent incomparable model groups from
+    being populated to the distance table.
+
+    Args:
+        train_end_times (list) The set of train_end_times to consider during
+                               the iteration
+        initial_model_group_ids (list) The initial list of model group ids to
+                                       narrow down
+        models_table (string) The name of the results schema
+        db_engine (sqlalchemy.engine) A database engine with access to results
+                                      schema of a completed modeling run
+    """
+
+    if isinstance(train_end_times, str) or not hasattr(train_end_times, '__iter__'):
+        raise TypeError("train_end_times should be a list of str or timestamp")
+
+    if not bool(train_end_times):
+        raise ValueError("train_end_times shouldn't be an empty list")
+
+    end_times_sql = "ARRAY[{}]".format(
+        ', '.join(
+            "'{}'".format(
+                end_time.strftime('%Y-%m-%d') if isinstance(end_time, datetime) else end_time
+            ) for end_time in train_end_times
+        )
+    )
+    logging.info("Checking if all model groups have the same train end times")
+    logging.info(f"Found {len(initial_model_group_ids)} total model groups")
+    query = f"""
+        SELECT model_group_id
+        FROM (
+            SELECT
+                model_group_id,
+                array_agg(distinct train_end_time::date::text) as train_end_time_list
+            FROM model_metadata.{models_table}
+            WHERE model_group_id in {tuple(initial_model_group_ids)}
+            GROUP BY model_group_id
+        ) as t
+        WHERE train_end_time_list = {end_times_sql}
+        """
```

Comment:
And, should we leverage the SQLAlchemy engine, we can relieve ourselves of having to properly interpolate values, (and of having to escape values against the unlikely possibility of a SQL injection attack or, more likely here, a crazy mistake).

So long as we're building our own SQL statement from strings, which is fine here, we can't get away from interpolating the table name; but, that's not really an issue. Rather, the danger is in interpolating values, (the right-hand side of comparative operations).

Above, we are relying on the `repr` of the Python `tuple` to be precisely the same as that of the database backend. That's actually *not* the case even in PostgreSQL, at least not for the single-element tuple.

In Python: `(item,)` –

    >>> assert 'jocodssg_read' in ('jocodssg_read',)

In PostgreSQL: `(item)` –

    postgres=> select rolname from pg_roles where rolname in ('jocodssg_read');
        rolname    
    ---------------
     jocodssg_read
    (1 row)

    postgres=> select rolname from pg_roles where rolname in ('jocodssg_read',);
    ERROR:  syntax error at or near ")"                                                
    LINE 1: ...rolname from pg_roles where rolname in ('jocodssg_read',);

Using SQLAlchemy `execute()`, we can just pass values to let the library bind them appropriately, *e.g.*, (sorry from memory):

    db_engine.execute(f"select col0 from {my_table} where col0 in :val0", val0=a_list_or_tuple)

SQLAlchemy will then figure out how to represent `val0` appropriately, (and protect against it being something insane).

It gets even better, IIRC, when you use `text`. I don't remember all the ins and outs, but for example, I believe this is the way you would hand over construction of your ARRAY to SQLAlchemy as well:

    from sqlalchemy import bindparam, text, String
    from sqlalchemy.dialects.postgresql import ARRAY

    end_time_values = [
        end_time.strftime('%Y-%m-%d') if isinstance(end_time, datetime) else end_time
        for end_time in end_times
    ]
    query = text(
        f'select col0 from {my_table} where col1 = :val0'
    ).bindparams(
        bindparam('val0', ARRAY(String)),
    )
    db_engine.execute(query, val0=end_time_values)